PD-1369 / 25.04 / Pd 1369 update custom app (#3123)
truenas/documentation: content/TruenasApps/UsingCustomApp.md

To create directories in a dataset on TrueNAS, before you begin installing the container, open the TrueNAS CLI and enter `storage filesystem mkdir path="/PATH/TO/DIRECTORY"`.
{{< /hint >}}

{{< include file="/static/includes/apps/AppsCustomApp.md" >}}

## Setting up App Storage

> 

Author: I did but in most cases they did not work as well for the general custom app process as they do for a given individual app

### Using ixVolumes

Using ixVolumes, you can allow TrueNAS to create a dataset on the apps storage pool.
Like host paths, ixVolumes are mounted as Docker bind mounts and can be mounted read only.
In most use cases, ixVolume storage is better suited to app testing, while host volumes are better for production deployments.

### Using SMB Shares

Users can mount a an SMB share for storage within the container using Docker [volumes](https://docs.docker.com/engine/storage/#volumes).
Volumes consume space from the pool chosen for application management.
You need to define a path where the volume appears inside the container.

### Using Memory-Backed Storage

Memory-backed [tmpfs](https://docs.docker.com/engine/storage/#tmpfs) storage provides temporary, in-memory filesystem for storing data that only needs to persist for the lifetime of the container.
It is mounted directly in RAM and automatically cleared when the container stops.

Memory-backed storage can be used to store ephemeral data, especially when performance speed is a concern, such as caching, processing data that doesn't need persistent storage, or storing session data.

> For security purposes, sensitive data like temporary credentials or data can be stored in tmpfs instead of writing to disk.

Author: I'd rather present it as an option than a suggestion. Revising wording a little

   Enter the **Port** number to use for portal access.
   The port number the app uses should be in the documentation provided by the application provider/developer.
   Check the port number against the list of [Default Ports](https://www.truenas.com/docs/references/defaultports/) to make sure TrueNAS is not using it for some other purpose.
   If needed, you can map the default container port to an open host port under **Network Configuration** (see above).

   If needed, enter a custom **Path** for portal access, for example */admin*. Defaults to */*. The path is appended to the host IP and port, as in **truenas.local:15000/admin**.

7. Add the **Storage** settings.
   Review the image documentation for required storage volumes.
   See [Setting up Storage](#setting-up-app-storage) above for more information.

   {{< trueimage src="/images/SCALE/Apps/InstallCustomAppScreenStorage.png" alt="Storage Settings" id="Storage Settings" >}}

   Click **Add** to display a block of configuration settings for each storage mount.
   You can edit to add additional storage to the container later, if needed.

>    Use the dropdown to select the storage **Type** you need to mount then configure the appropriate settings.

Reviewer: ```suggestion
   Select the storage option you need to mount  on the **Type** dropdown list and then configure the appropriate settings.
```

   {{< /expand >}}

   {{< expand "Host Path (Path that already exists on the system)" "v" >}}
   To configure a persistent host path to an existing dataset, select **Host Path (Path that already exists on the system)**.
   This is the recommended setting for on-disk storage in a production deployment.

   {{< trueimage src="/images/SCALE/Apps/InstallCustomAppAddHostPathVol.png" alt="Host Path Settings" id="Host Path Settings" >}}

   Enter in **Mount Path** the <file>**path/to/directory**</file> where the host path mounts inside the container.
   Then define the **Host Path**. Enter a path or click <span class="material-icons">arrow_right</span> to the left of <svg xmlns="http://www.w3.org/2000/svg" width="1em" height="1em" viewBox="0 0 24 24"><path fill="currentColor" d="M5 21L3 9h18l-2 12zm5-6h4q.425 0 .713-.288T15 14t-.288-.712T14 13h-4q-.425 0-.712.288T9 14t.288.713T10 15M6 8q-.425 0-.712-.288T5 7t.288-.712T6 6h12q.425 0 .713.288T19 7t-.288.713T18 8zm2-3q-.425 0-.712-.288T7 4t.288-.712T8 3h8q.425 0 .713.288T17 4t-.288.713T16 5z"/></svg> **/mnt** to browse to the location of the dataset in TrueNAS.

>    Select whether to mount the host path as **Read Only** or to **Enable ACL** and configure settings as needed.

Reviewer: in most of the apps, you need to enable the acl or the app does not deploy, so is that the case here as well? Also, you could use the snippets created for apps tutorials for this content.

Author: My understanding of it is you need to set permissions correctly, but that could be directly on the dataset or by enabling ACL here? 

And this is one of the cases where the snippet doesn't work, because the ACL snippet includes the sentence "* The default apps user which is either **473** for apps in the stable train or **568** for apps in the **enterprise** or **community** train," which doesn't make sense in the Custom App process. 

I can bring over a bit more detail from the snippet but it would read awkwardly to just include it.

Reviewer: On ACLs, I asked Stavros if you could use the dataset permissions widget to access the Edit ACL form and he said no, you must use the ACL option in the container through the install widget.
If Stavros is not imposing this ACL requirement in the custom apps wizard then we just need to be clear that it is an option but not a requirement.

Author: Checking with Stavros.

   {{< /expand >}}

   {{< expand "SMB/CIFS Share (Mounts a volume to a SMB share)" "v" >}}
   To mount an existing SMB share as a Docker [volume](https://docs.docker.com/engine/storage/#volumes), select **SMB/CIFS Share (Mounts a volume to a SMB share)**.

   {{< trueimage src="/images/SCALE/Apps/InstallCustomAppAddSMB.png" alt="SMB/CIFS Share Settings" id="SMB/CIFS Share Settings" >}}

>    Enter in **Mount Path** the <file>**path/to/directory**</file> where the ixVolume mounts inside the container.

Reviewer: ```suggestion
   Enter the <file>**path/to/directory**</file> where the ixVolume mounts inside the container in **Mount Path**.
```

>    Enter in **Server** the IP address for the SMB server, for example *192.168.1.100*.
>    This can be the TrueNAS host.
> 
>    Enter the name of the SMB share in **Path**, for example *my-share*.
> 
>    Enter the **Username** and **Password** of an account with permission to access the SMB share.

Reviewer: In the use case in these sentences we should not make the field the noun and use UI emphasis. Suggest the following changes:
```suggestion
   Enter the IP address for the SMB server in **Server**, for example *192.168.1.100*.
   This can be the TrueNAS host.

   Enter the name of the SMB share in **Path**, for example *my-share*.

   Enter the username and password of an account with permission to access the SMB share.
```

   Select whether to mount the share volume as **Read Only** or enter a directory services **Domain** as needed.
   {{< /expand >}}

   {{< expand "Tmpfs (Temporary directory created on the RAM)" "v" >}}
   To mount a temporary memory-backed storage directory, select **Tmpfs (Temporary directory created on the RAM)**.

   {{< trueimage src="/images/SCALE/Apps/InstallCustomAppAddMemoryBackedVol.png" alt="Tmpfs Settings" id="Tmpfs Settings" >}}

>    Enter in **Mount Path** the <file>**path/to/directory**</file> where the temporary directory mounts inside the container.

Reviewer: ```suggestion
   Enter the <file>**path/to/directory**</file> where the temporary directory mounts inside the container in **Mount Path**.
```

>    Enter in **Tmpfs Size Limit (in Mi)** the maximum size of the temporary directory in mebibytes or accept the default of *500*.

Reviewer: ```suggestion
   Enter the maximum size of the temporary directory in mebibytes in **Tmpfs Size Limit (in Mi)** or accept the default of *500*.
```

   {{< /expand >}}

8. {{< include file="/static/includes/apps/AppInstallWizardResourceConfig.md" >}}

9. Use **GPU Configuration** to allocate GPU resources if available and required for the application.

10. Click **Install** to deploy the container.
    If you correctly configured the app, the widget displays on the **Installed Applications** screen.

    When complete, the container becomes active.
    If the container does not automatically start, click **Start** on the widget.

    Click on the App card reveals details.

## Installing via YAML

{{< include file="/static/includes/apps/YAMLWarning.md" >}}

To install a third-party application using a Docker Compose YAML file, go to **Apps > Discover**.
Click <i class="material-icons" aria-hidden="true" title="more_vert">more_vert</i> then select **Install via YAML** to open the **Add Custom App** screen.

{{< trueimage src="/images/SCALE/Apps/InstallCustomAppYAML.png" alt="Install Custom App via YAML" id="Install Custom App via YAML" >}}

> 
> The **Add Custom App** screen contains two fields.

Reviewer: This sentence is not needed here, suggest removing it.
```suggestion

```

> Enter a name for the application in **Name** using lowercase, alphanumeric characters.

Reviewer: i don't think you need to include the field name here
```suggestion
Enter a name for the application using lowercase, alphanumeric characters.
```

Enter the Compose YAML file in **Custom Config**.

> Ensure the Compose file begins with a top-level element, such as `name:` or `services:`.

Reviewer: I hate when we say "ensure". If you are telling them what to do here, which you are, you don't need to say ensure. Thoughts?
```suggestion
Begin the Compose file with a top-level element, such as `name:` or `services:`.
```

See the Docker [Compose file reference](https://docs.docker.com/reference/compose-file/) for more information.

As Compose files can be complex and YAML relies on indentation and whitespace to define structure and hierarchy, we recommend writing the file in a stand-alone code editor before pasting the completed content into the **Custom Config** field.

Click **Save** to begin deployment of the app.

### YAML Deployment Examples

The following examples represent some of the capabilities of Docker Compose and how it can be used to configure applications in TrueNAS.
This is not an exhaustive collection of all capabilities, nor are the files below intended as production-ready deployment templates.

> Note that app storage in these examples is configured using [volumes](https://docs.docker.com/engine/storage/#volumes), which are managed by the Docker engine.

Reviewer: ```suggestion
Note that app storage in these examples are configured using [volumes](https://docs.docker.com/engine/storage/#volumes), which are managed by the Docker engine.
```

Author: The noun here is "storage", so "is" agrees with the noun.

> In a production deployment, you should configure [bind mounts](https://docs.docker.com/engine/storage/#bind-mounts) to TrueNAS storage locations.

Reviewer: ```suggestion
In a production deployment, configure [bind mounts](https://docs.docker.com/engine/storage/#bind-mounts) to TrueNAS storage locations.
```

        volume: {}
      - type: volume
        source: thumbs
        target: /usr/src/app/upload/thumbs
        volume: {}
      - type: volume
        source: profile
        target: /usr/src/app/upload/profile
        volume: {}
      - type: volume
        source: video
        target: /usr/src/app/upload/encoded-video
        volume: {}
      - type: volume
        target: /tmp
        volume: {}
      - type: volume
        target: /scratchpad
        volume:
          nocopy: true
networks:
  default:
    name: ad7a9839cdf6cde9e34b526956ba5f37_default
volumes:
  library:
    name: ad7a9839cdf6cde9e34b526956ba5f37_library
  postgres-data:
    name: ad7a9839cdf6cde9e34b526956ba5f37_postgres-data
  profile:
    name: ad7a9839cdf6cde9e34b526956ba5f37_profile
  thumbs:
    name: ad7a9839cdf6cde9e34b526956ba5f37_thumbs
  tmp:
    name: ad7a9839cdf6cde9e34b526956ba5f37_tmp
  uploads:
    name: ad7a9839cdf6cde9e34b526956ba5f37_uploads
  video:
    name: ad7a9839cdf6cde9e34b526956ba5f37_video{{< /codeblock >}}

#### Installing a Multi-Container App with a Dockerfile
This example deploys Nextcloud with multiple containers, including redis and postgres.

> A [Dockerfile](https://docs.docker.com/reference/dockerfile/) builds the Nextcloud image locally, rather than pulling from a remote repository.
> You can use an existing Dockerfile or, as in the example, use [`dockerfile_inline`](https://docs.docker.com/reference/compose-file/build/#dockerfile_inline) to define the build commands as an inline string within the Compose file.

Reviewer: I don't know that we need to capitalize dockerfile and compose in these sentences as we are not taking about proper nouns. Thoughts?
```suggestion
A [dockerfile](https://docs.docker.com/reference/dockerfile/) builds the Nextcloud image locally, rather than pulling from a remote repository.
You can use an existing dockerfile or, as in the example, use [`dockerfile_inline`](https://docs.docker.com/reference/compose-file/build/#dockerfile_inline) to define the build commands as an inline string within the compose file.
```

Author: They are proper nouns though. Dockerfile is the name of a specific thing. Compose is being used as the name of the [Docker] Compose software, not "compose," which is a verb.

A Google search shows most other online docs (Docker or otherwise) are capitalizing Dockerfile

      set -e
      config_file="/var/www/html/config/config.php"

      echo "Updating database and redis host in config.php"
      sed -i "s/\('dbhost' => '\)[^']*postgres:5432',/\1postgres:5432',/" "$$config_file"
      occ config:system:set redis host --value="redis"
  limitrequestbody.conf:
    name: ade25cf7c52e909aeef17c9aaee373aa_limitrequestbody.conf
    content: |2
      LimitRequestBody 3221225472
  occ:
    name: ade25cf7c52e909aeef17c9aaee373aa_occ
    content: |
      #!/bin/bash
      uid="$$(id -u)"
      gid="$$(id -g)"
      if [ "$$uid" = '0' ]; then
        user='www-data'
        group='www-data'
      else
        user="$$uid"
        group="$$gid"
      fi
      run_as() {
        if [ "$$(id -u)" = 0 ]; then
          su -p "$$user" -s /bin/bash -c "php /var/www/html/occ $$(printf '%q ' "$$@")"
        else
          /bin/bash -c "php /var/www/html/occ $$(printf '%q ' "$$@")"
        fi
      }
      run_as "$$@"
  opcache.ini:
    name: ade25cf7c52e909aeef17c9aaee373aa_opcache.ini
    content: |
      opcache.memory_consumption=128
  php.ini:
    name: ade25cf7c52e909aeef17c9aaee373aa_php.ini
    content: |
      max_execution_time=30{{< /codeblock >}}

> Alternatively, you can use [`dockerfile`](https://docs.docker.com/reference/compose-file/build/#dockerfile) to set an existing Dockerfile stored on your TrueNAS system.
> 
> To specify the absolute path to a Dockerfile, remove this `dockerfile_inline` section of the `build` element:

Reviewer: same comment about capitalzing dokerfile as it is not a proper noun.
```suggestion
Alternatively, you can use [`dockerfile`](https://docs.docker.com/reference/compose-file/build/#dockerfile) to set an existing dockerfile stored on your TrueNAS system.

To specify the absolute path to a dockerfile, remove this `dockerfile_inline` section of the `build` element:
```

Author: same reply

{{< codeblock language="yaml" >}}nextcloud:
    build:
      dockerfile_inline: |-
        FROM nextcloud:30.0.0
        RUN apt update || { echo "Failed to update apt cache. Exiting."; exit 1; }
        RUN apt install -y --no-install-recommends ffmpeg || { echo "Failed to install [ffmpeg]. Exiting."; exit 1; }
        RUN apt install -y --no-install-recommends smbclient || { echo "Failed to install [smbclient]. Exiting."; exit 1; }
        RUN apt install -y --no-install-recommends ocrmypdf || { echo "Failed to install [ocrmypdf]. Exiting."; exit 1; }
        RUN apt install -y --no-install-recommends libsmbclient-dev
        RUN pecl install smbclient
        RUN docker-php-ext-enable smbclient
        RUN ldd "$$(php -r 'echo ini_get("extension_dir");')"/smbclient.so
        RUN apt install -y --no-install-recommends tesseract-ocr-eng || { echo "Failed to install [tesseract-ocr-eng]. Exiting."; exit 1; }
        RUN apt install -y --no-install-recommends tesseract-ocr-chi-sim || { echo "Failed to install [tesseract-ocr-chi-sim]. Exiting."; exit 1; }{{< /codeblock >}}

> Replace it with the absolute path to the Dockerfile, for example:

Reviewer: ```suggestion
Replace it with the absolute path to the dockerfile, for example:
```

      - /mnt/virtual/library/tv:/tv:ro #UPDATE LOCAL PATH
      - /mnt/virtual/library/movies:/movies:ro #UPDATE LOCAL PATH
    tmpfs:
      - /transcode
    healthcheck:
      test: 'curl --silent --fail http://localhost:32400/identity'
      interval: 15s
      timeout: 5s
      retries: 15
      start_period: 60s{{< /codeblock >}}

**Tautulli**

{{< codeblock language="yaml" >}}networks:
  plex_net:
    name: plex_net
    driver: bridge
    external: true

services:
  tautulli:
    image: tautulli/tautulli:v2.14.5@sha256:6017b491d8e9100a97391b639fff5824ad36a315c69aae3c9ed78407994a626e
    restart: always
    entrypoint:
      - python
      - Tautulli.py
    command:
      - --port
      - "8989"
      - --config
      - /config/config.ini
      - --datadir
      - /config
    networks:
      - plex_net
    volumes:
      - /mnt/cache/appdata/tautulli:/config
      - /mnt/cache/appdata/plex/Library/Application Support/Plex Media Server/Logs:/plexlogs:ro
    ports:
      8989:8989
    healthcheck:
      test: curl --silent --fail http://localhost:8989/status
      interval: 10s
      timeout: 5s
      retries: 5
      start_period: 30s{{< /codeblock >}}

## Managing Custom Apps

> Installed custom applications appear on the **Installed** applications screen.

Reviewer: ```suggestion
Installed custom applications show on the **Installed** applications screen.
```
